File: Sample/Vibes SDK Objective-C Sample App/InboxViewController.m
//
//  InboxViewController.m
//  Vibes SDK Objective-C Sample App
//
//  Created by Moin' Victor on 24/11/2019.
//  Copyright © 2019 Vibes. All rights reserved.
//

#import "InboxViewController.h"
#import "InboxMessageTableViewCell.h"
#import "InboxDetailViewController.h"
@import VibesPush;

@interface InboxViewController ()

@end

@implementation InboxViewController
    NSArray<InboxMessage *> *inboxMessages;
    NSInteger selectedIndex;
- (void)viewDidLoad {
    [super viewDidLoad];
    inboxMessages = [[NSMutableArray alloc] init];
    self.refreshControl = [[UIRefreshControl alloc] init];
    [self.refreshControl addTarget:self action:@selector(refreshTable) forControlEvents: UIControlEventValueChanged];

    if (@available(iOS 10.0, *)) {
        self.tableView.refreshControl = self.refreshControl;
    } else {
        [self.tableView addSubview: self.refreshControl];
    }
    [self.tableView setDelegate:self];
    [self.tableView setDataSource:self];
    // Do any additional setup after loading the view.
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self fetchMessages];
}

- (void)refreshTable {
    //TODO: refresh your data
    [self fetchMessages];
}

- (void) fetchMessages {
    [self.refreshControl beginRefreshing];
    [[Vibes shared] fetchInboxMessages:^(NSArray<InboxMessage *> * _Nonnull messages, NSError * _Nullable error) {
        NSLog(@"fetchInboxMessages: Result -----===>>>>>>> %@", messages);
        NSLog(@"fetchInboxMessages: Error -----===>>>>>>> %@", error ? error : @"[None]");
        dispatch_async(dispatch_get_main_queue(), ^(void) {
            [self.refreshControl endRefreshing];
            // Run UI Updates
            if (error == nil) {
                inboxMessages = messages;
                [self.tableView reloadData];
            } else {
                UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Inbox Error" message: error.localizedDescription preferredStyle:UIAlertControllerStyleAlert];

                UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"Dismiss" style:UIAlertActionStyleCancel handler:nil];
                [alert addAction: cancel];
                [self presentViewController: alert animated:YES completion:nil];
            }
        });
    }];
}

#pragma mark - Table View Data source
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:
(NSInteger)section
{
    return [inboxMessages count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:
(NSIndexPath *)indexPath
{
    static NSString *cellId = @"Cell";

    InboxMessageTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:
                             cellId];
    if (cell == nil) {
        cell = [[InboxMessageTableViewCell alloc]initWithStyle:
                UITableViewCellStyleDefault reuseIdentifier: cellId];
    }
    InboxMessage *message = [inboxMessages objectAtIndex: indexPath.row];

    [cell.contentLabel setText: message.content];
    [cell.subjectLabel setText: message.subject];
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    dateFormatter.dateFormat = @"yyyy-MM-dd 'at' HH:mm:ss";
    NSString *dateString = [dateFormatter stringFromDate: message.createdAt];
    [cell.dateLabel setText:dateString];

    [self downloadImageFromURL:message.detail intoImageView:cell.imageIcon withActivityIndicator:cell.activityIndicator];
    return cell;
}

// Default is 1 if not implemented
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    if (inboxMessages == nil || inboxMessages.count == 0) {
        UILabel *noDataLabel = [[UILabel alloc]initWithFrame:CGRectMake(0, 0, self.tableView.bounds.size.width, tableView.bounds.size.height)];
        [noDataLabel setText:@"No inbox messages. Pull to fetch messages... "];
        [noDataLabel setTextColor: [UIColor grayColor]];
        [noDataLabel setTextAlignment: NSTextAlignmentCenter];
        tableView.backgroundView = noDataLabel;
        [tableView setSeparatorStyle: UITableViewCellSeparatorStyleNone];
        return 0;
    }
    self.tableView.backgroundView = nil;
    [tableView setSeparatorStyle:UITableViewCellSeparatorStyleSingleLine];
    return 1;
}

#pragma mark - TableView delegate

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:
(NSIndexPath *)indexPath{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    InboxMessageTableViewCell *cell = [tableView cellForRowAtIndexPath:indexPath];
    NSLog(@"Row:%d selected and its data is %@", indexPath.row, cell.subjectLabel.text);
    selectedIndex = indexPath.row;
    [self performSegueWithIdentifier:@"inbox_detail" sender:nil];
}

#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
    if ([segue.identifier isEqualToString: @"inbox_detail"]) {
        InboxDetailViewController *detailCont = segue.destinationViewController;
        detailCont.inboxMessage = [inboxMessages objectAtIndex:selectedIndex];
    }
}

/// Download the image into image view
/// @param url Image URL
/// @param imageView The image view
/// @param activityIndicator The activity indicator to show progress
- (void) downloadImageFromURL: (NSString*) url intoImageView: (UIImageView*) imageView
        withActivityIndicator: (UIActivityIndicatorView*) activityIndicator
{
    [activityIndicator startAnimating];
    [activityIndicator setHidden:NO];
    dispatch_async(dispatch_get_global_queue(0,0), ^{
        NSData * data = [[NSData alloc] initWithContentsOfURL: [NSURL URLWithString: url]];
        if ( data == nil) {
            dispatch_async(dispatch_get_main_queue(), ^{
                [activityIndicator stopAnimating];
                [activityIndicator setHidden:YES];
            });
            return;
        }
        dispatch_async(dispatch_get_main_queue(), ^{
            // WARNING: is the cell still using the same data by this point??
            imageView.image = [UIImage imageWithData: data];
            [activityIndicator stopAnimating];
            [activityIndicator setHidden:YES];
        });
    });
}

@end

File: SampleSwift/VibesSDKSampleApp/InboxViewController.swift
import UIKit
import VibesPush

/// Inbox Tab
class InboxViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
    var inboxMessages: [InboxMessage] = []

    var selectedItem: Int?
    private let refreshControl = UIRefreshControl()
    @IBOutlet var tableView: UITableView!

    override func viewDidLoad() {
        super.viewDidLoad()
        // Configure Refresh Control
        refreshControl.addTarget(self, action: #selector(refreshInboxData(_:)), for: .valueChanged)
        // Add Refresh Control to Table View
        if #available(iOS 10.0, *) {
            tableView.refreshControl = refreshControl
        } else {
            tableView.addSubview(refreshControl)
        }
        // remove table view empty cells
        tableView.tableFooterView = UIView()
        // Do any additional setup after loading the view.
    }

    @objc private func refreshInboxData(_ sender: Any) {
        // Fetch inbox messages
        fetchInboxMessages()
    }

    func fetchInboxMessages() {
        self.refreshControl.beginRefreshing()
        Vibes.shared.fetchInboxMessages { [weak self] inboxMessages, error in
            DispatchQueue.main.async {
                self?.refreshControl.endRefreshing()
                print("fetchInboxMessages: Result -----===>>>>>>>", inboxMessages)
                print("fetchInboxMessages: Error -----===>>>>>>>", error ?? "[None]")
                if error == nil {
                    self?.inboxMessages = inboxMessages
                    self?.tableView.reloadData()
                } else {
                    // show the error message
                    let alertController = UIAlertController(title: "Inbox Error", message: error?.localizedDescription, preferredStyle: .alert)
                    alertController.addAction(UIAlertAction(title: "Dismiss", style: .default, handler: { (action) in
                        self?.refreshControl.endRefreshing()
                    }))
                    self?.present(alertController, animated: true, completion: nil)
                }
            }
        }
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        fetchInboxMessages()
    }

    func numberOfSections(in tableView: UITableView) -> Int {
        // handle empty state, show message when no data is available
        var numOfSections: Int = 0
        if inboxMessages.count > 0 {
            tableView.separatorStyle = .singleLine
            numOfSections = 1
            tableView.backgroundView = nil
        } else {
            let noDataLabel: UILabel = UILabel(frame: CGRect(x: 0, y: 0, width: tableView.bounds.size.width, height: tableView.bounds.size.height))
            noDataLabel.text = "No inbox messages. Pull to fetch messages... "
            noDataLabel.textColor = UIColor.gray
            noDataLabel.textAlignment = .center
            tableView.backgroundView = noDataLabel
            tableView.separatorStyle = .none
        }
        return numOfSections
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return inboxMessages.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let inboxCell = tableView.dequeueReusableCell(withIdentifier: "InboxCell", for: indexPath) as! InboxTableViewCell
        let inboxMessage = self.inboxMessages[indexPath.row]
        print("------>>>>>>>>", inboxMessages)
        inboxCell.subjectLabel.text = inboxMessage.subject
        inboxCell.contentLabel.text = inboxMessage.content
        let date = inboxMessage.createdAt
        let dateFormat = DateFormatter()
        dateFormat.dateFormat = "yyyy-MM-dd hh:mm:ss"
        inboxCell.dateLabel.text = dateFormat.string(from: date)
        if let detail = inboxMessage.detail {
            inboxCell.imageWidth.constant = 52
            downloadImage(from: URL(string: detail)!, into: inboxCell.imageIcon, withIndicator: inboxCell.activityIndicator)
        } else {
            inboxCell.imageWidth.constant = 0
        }

        return inboxCell
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        selectedItem = indexPath.row
        performSegue(withIdentifier: "inbox_detail", sender: nil)
        tableView.deselectRow(at: indexPath, animated: true)
    }
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
        if segue.identifier == "inbox_detail",
            let selectedItem = selectedItem,
            let controller = segue.destination as? InboxDetailViewController {
            controller.inboxMessage = inboxMessages[selectedItem]
        }
    }
}

// Mark: UIViewController Image loading
extension UIViewController {

    /// Get Data
    /// - Parameters:
    ///   - url: The url
    ///   - completion: Completion callbacl
    func getData(from url: URL, completion: @escaping (Data?, URLResponse?, Error?) -> ()) {
        URLSession.shared.dataTask(with: url, completionHandler: completion).resume()
    }

    /// Download image into URL into image view
    /// - Parameters:
    ///   - url: The url
    ///   - imageView: The image view
    ///   - activityIndicator: Activity indicator to show progress
    func downloadImage(from url: URL, into imageView: UIImageView?, withIndicator activityIndicator: UIActivityIndicatorView?) {
        print("Download Started")
        activityIndicator?.isHidden = false
        activityIndicator?.startAnimating()
        getData(from: url) { data, response, error in
            guard let data = data, error == nil else { return }
            print(response?.suggestedFilename ?? url.lastPathComponent)
            print("Download Finished")
            DispatchQueue.main.async() {
                activityIndicator?.stopAnimating()
                activityIndicator?.isHidden = true
                imageView?.image = UIImage(data: data)
            }
        }
    }
}
